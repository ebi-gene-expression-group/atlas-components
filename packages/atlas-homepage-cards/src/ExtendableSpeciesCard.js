import React from 'react'
import PropTypes from 'prop-types'
import EbiSpeciesIcon from 'react-ebi-species'

const MAX = 5

class ExtendableSpeciesCard extends React.Component {

  constructor(props) {
    super(props)

    this.state = {
      isHidden: this.props.content && this.props.content.length >= MAX
    }
    this.onClick = this.onClick.bind(this)
  }
 
  onClick() {
    this.setState({
      isHidden: !this.state.isHidden
    })
  }
    
  render() {
    const {iconSrc, iconDescription, content} = this.props

    const visibleContent = content && content.map((item) => {
      return item.url ?
        <li className={`url`} style={{marginBottom: `0.3rem`}} key={item.text}><a href={item.url} key={item.text}>{item.text}</a></li> :
        <li className={`text`} style={{marginBottom: `0.3rem`}} key={item.text}>{item.text}</li>
    })

    return (
      <div className={`column column-block text-center combo card`} style={{marginBottom:0, paddingBottom: `25px`}}>
        {
          iconDescription && <h5 className={`species-name`}>{iconDescription}</h5>
        }

        <span className={`species-icon`} style={{fontSize: `800%`}}>
          <EbiSpeciesIcon species={iconSrc}/>
        </span>
        {
          content &&
            <ul className={`content`} style={{listStyle: `none`, paddingLeft: `0`, marginLeft: `0`}}>
              <div>
              {
                this.state.isHidden ?
                  visibleContent.slice(0, MAX) :
                  visibleContent
              }
              {
                content.length > MAX &&
                <button className={`button small`} onClick={this.onClick}>{this.state.isHidden ? `Show all` : `Show less`}</button>
              }  
              </div>
            </ul>
        }
      </div>
    )
  }
}

ExtendableSpeciesCard.propTypes = {
  iconSrc: PropTypes.string.isRequired,
  iconDescription: PropTypes.string,
  content: PropTypes.arrayOf(PropTypes.shape({
    text: PropTypes.string.isRequired,
    url: PropTypes.string
  }))
}

export default ExtendableSpeciesCard
