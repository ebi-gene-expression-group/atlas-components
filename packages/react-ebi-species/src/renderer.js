"use strict";

//*------------------------------------------------------------------*

var React = require('react');
var ReactDOM = require('react-dom');

//*------------------------------------------------------------------*

var Icon = require('./SpeciesIcon.jsx');

//*------------------------------------------------------------------*

module.exports = function(mountNode,species,colourOverride, colourPerKingdom) {
    ReactDOM.render(
        React.createElement(Icon,{
          species: species,
          colourOverride: colourOverride,
          colourPerKingdom: colourPerKingdom
        }), mountNode
    );
};
