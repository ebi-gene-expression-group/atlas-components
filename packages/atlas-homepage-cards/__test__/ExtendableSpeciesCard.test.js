import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import {shallow} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import EbiSpeciesIcon from 'react-ebi-species'
import ExtendableSpeciesCard from '../src/ExtendableSpeciesCard.js'

Enzyme.configure({ adapter: new Adapter() })

describe(`ExtendableSpeciesCard`, () => {
  test(`with no data matches snapshot`, () => {
    const tree = renderer.create(<ExtendableSpeciesCard iconSrc={``}/>).toJSON()
    expect(tree).toMatchSnapshot()
  })


  test(`does not render optional empty content`, () => {
    const wrapper = shallow(<ExtendableSpeciesCard iconSrc={``} />)

    expect(wrapper.find(EbiSpeciesIcon).exists()).toBe(true)
    expect(wrapper.find(`.species-name`).exists()).toBe(false)
    expect(wrapper.find(`.content`).exists()).toBe(false)
  })

  test(`does not render non-existent descriptions`, () => {
    const content = [
      {
        text: `120 experiments`
      },
      {
        text: `1 baseline experiment`
      },
      {
        text: `119 differential experiments`
      }
    ]
    const speciesName = `Mus musculus`

    const wrapper = shallow(<ExtendableSpeciesCard iconSrc={speciesName} content={content} description={speciesName}/>)
    expect(wrapper.find(`.species-name`).exists()).toBe(false)

    const contentWrapper = wrapper.find(`.content`)
    expect(contentWrapper.exists()).toBe(true)
    expect(contentWrapper.find(`.content .text`)).toHaveLength(3)
    expect(contentWrapper.find(`.content .url`).exists()).toBe(false)
  })

  test(`renders URLs`, () => {
    const content = [
      {
        text: `120 experiments`,
        url: `http://example.com/experiments`
      }
    ]

    const wrapper = shallow(<ExtendableSpeciesCard iconSrc={``} content={content}/>)
    const urlsWrapper = wrapper.find(`.content .url`)

    expect(urlsWrapper).toHaveLength(1)
  })

  test(`displays upto five experiments be default`, () => {
    const content = [
      {
        text: `1 experiments`
      },
      {
        text: `2 baseline experiment`
      },
      {
        text: `3 differential experiments`
      },
      {
        text: `4 differential experiments`
      },
      {
        text: `5 differential experiments`
      },
      {
        text: `6 differential experiments`
      }
    ]
    const speciesName = `Mus musculus`

    const wrapper = shallow(<ExtendableSpeciesCard iconSrc={speciesName} content={content} description={speciesName}/>)
    expect(wrapper.find(`.content .text`)).toHaveLength(5)
    wrapper.setState({ isHidden: false })
    wrapper.update()
    expect(wrapper.find(`.content .text`)).toHaveLength(content.length)
  })
})
