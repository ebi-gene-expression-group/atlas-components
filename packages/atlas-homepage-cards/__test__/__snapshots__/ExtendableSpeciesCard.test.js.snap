// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtendableSpeciesCard with no data matches snapshot 1`] = `
<div
  className="column column-block text-center combo card"
  style={
    Object {
      "marginBottom": 0,
      "paddingBottom": "25px",
    }
  }
>
  <span
    className="species-icon"
    style={
      Object {
        "fontSize": "800%",
      }
    }
  >
    <span
      className="icon icon-species"
      data-icon="❔"
      style={
        Object {
          "color": "black",
        }
      }
      title=""
    />
  </span>
</div>
`;
